runtime/metricsview/ast.go

```diff
@@ -1101,6 +1148,53 @@ func (a *AST) sqlForAnyInGroup(expr string) string {
 	return fmt.Sprintf("ANY_VALUE(%s)", expr)
 }
 
+// sqlForExpression returns the provided time expression adjusted by the fixed time offset between the current query's base and comparison time ranges.
+func (a *AST) sqlForExpressionAdjustedByComparisonTimeRangeOffset(expr string, g, mg TimeGrain) (string, error) {
+	if a.query.TimeRange == nil || a.query.TimeRange.Start.IsZero() || a.query.ComparisonTimeRange == nil || a.query.ComparisonTimeRange.Start.IsZero() {
+		return "", errors.New("must specify an explicit start time for both the base and comparison time range when comparing by a time dimension")
+	}
+
+	start1 := a.query.TimeRange.Start
+	start2 := a.query.ComparisonTimeRange.Start
+
+	var dateDiff string
+	if g == TimeGrainUnspecified {
+		g = TimeGrainMillisecond // todo millis won't work for druid
+		res, err := a.dialect.DateDiff(g.ToProto(), start1, start2)
+		if err != nil {
+			return "", err
+		}
+		dateDiff = res
+	} else if g == mg {
+		res, err := a.dialect.DateDiff(g.ToProto(), start1, start2)
+		if err != nil {
+			return "", err
+		}
+		dateDiff = res
+	} else {
+		// larger time grain values can change as well
+		res, err := a.dialect.DateDiff(mg.ToProto(), start1, start2)
+		if err != nil {
+			return "", err
+		}
+		dateDiff = res
+
+		// DATE_TRUNC('year', t - INTERVAL (DATE_DIFF(start, end)) day)
+		tc := a.dialect.EscapeIdentifier(a.metricsView.TimeDimension)
+		expr := fmt.Sprintf("(%s - INTERVAL (%s) %s)", tc, dateDiff, a.dialect.ConvertToDateTruncSpecifier(mg.ToProto()))
+		dim := &runtimev1.MetricsViewSpec_DimensionV2{
+			Expression: expr,
+		}
+		expr, err = a.dialect.DateTruncExpr(dim, g.ToProto(), a.query.TimeZone, int(a.metricsView.FirstDayOfWeek), int(a.metricsView.FirstMonthOfYear))
+		if err != nil {
+			return "", fmt.Errorf(`failed to compute time floor: %w`, err)
+		}
+		return expr, nil
+	}
```

Comment:
This case completely discards the input `expr`, and changes to using `a.metricsView.TimeDimension`. What if the input `expr` was not that of the `a.metricsView.TimeDimension` (like when we support multiple time dimensions, as previously discussed)? This could be fixed by moving the handling to `resolveDim` and returning a `comparisonExpression` from there or something like that. Or at least there should be a `// TODO` here explaining the inconsistency.

Reply:
I expect `a.metricsView.TimeDimension` contains the required time column from a list of possible time columns.
From that point of view it will work.

Reply:
I think at least a comment about that would still be nice – it's pretty convoluted logic

Reply:
Done.